src/Network/MQTT/Encoder.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE OverloadedStrings, RecordWildCards #-}
+module Network.MQTT.Encoder (encodePacket) where
+
+import           Data.Bits (shiftL, (.|.))
+import           Data.ByteString.Builder
+import           Data.Foldable()
+import           Data.Int (Int64)
+import           Data.List
+import           Data.Maybe (isJust)
+import           Data.Monoid ((<>))
+import           Data.Text (Text)
+import           Data.Text.Encoding (encodeUtf8)
+import           Data.Word (Word8)
+import           Network.MQTT.Packet
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+
+encodePacket :: Packet -> Builder
+encodePacket packet = mconcat [fixedHeader', variableHeader', payload']
+  where
+    variableHeader' = variableHeader packet
+    payload' = payload packet
+    remaining = BSL.length $ toLazyByteString (variableHeader' <> payload')
+    fixedHeader' = fixedHeader packet remaining
```

Comment:
I would move `fixedHeader'` to the top of the where list so that the order will match with the order in mconcat.